### ISSUE TITLE
Remove old Metacat API and associated tables

### ISSUE THREAD

**Issue by mbjones** [NCEAS/metacat#1365]
The old Metacat API has been deprecated for a couple of years, and it is time to remove it.  This should eliminate a bunch of the tables that are no longer needed, and therefore eliminate cruft and allow for faster INSERT operations because we no longer will need the xml PathQuery related tables.  Let's develop a list of the tables to be deleted here, which will give a good indication of which CRUD and Query functions will need to be updated to remove associated code. At a minimum, we probably want to remove:

- xml_index
- xml_returnfield
- xml_queryresult
- xml_path_index

And we also may be able to remove:

- xml_nodes
- xml_nodes_revisions
- xml_relation

The challenge in doing this is that we break backwards compatibility for earlier clients, particularly Morpho, but also possibly the admin API.  We need to consider the impact of this change on clients, but the 3.x change is the best place to introduce this major compatibility issue, and it will significant;y reduce our maintenance overhead through lower code complexity.

**Comment by laurenwalker** [comment]
Is it possible to keep support for `metacat?action=reindex` and `metacat?action=reindexall`?

**Comment by taojing2002** [comment]
Yep. but we will move them to the admin servlet.

**Comment by csjx** [comment]
We also need to be aware that CN installs from scratch require the `action=insert` method for initially loading the object format list.  We can of course do this at the PostgreSQL level, but our scripts currently use the old Metacat API. Same goes for adding object formats now.

**Comment by taojing2002** [comment]
The reason we need the metacat api to insert/update object format list 
is we need to know if the the given object is metadata/data in the 
CN.create method. Maybe we can modify the CN.create code to give a 
special case for the format object list file - don't need to determine 
the data object format to this namespace.  After doing this , we can use 
the MN.create to insert the the object format list from the scratch 
installation.  Keeping "action=insert" sounds too much.

**Comment by mbjones** [comment]
Is this just for bootstrapping a CN with the formats doc?  It seems that we should be able to find a way around that issue, especially given that the current suite of CNs already have one so its no longer needed, and we should be using the DataONE API for inserts anyways.  If the problem is that `create()` needs the list to do its thing, maybe we could default the CN to use `DATA` and `appplication/octet-stream` as the type for the first formats doc, which we can then change in the sysmeta once the format list is in place?

**Comment by taojing2002** [comment]
Yes, it is for bootstrapping a CN with the formats doc. Yes, I think we 
can find a way around that issue and we should use the CN.create to 
insert it.

**Comment by taojing2002** [comment]
Since we are going to use the hash store, I think we can remove `xml_documents` and `xml_revisions`

**Comment by mbjones** [comment]
Jing, don't we still actively use several columns from those tables for key metadata and some queries, including `user_owner`, `user_updated`, `date_created`, `date_updated`, and `server_location`? Seems like they were our primary tables in the original model, and lots of other tables linked through them.

**Comment by mbjones** [comment]
yeah, that overlap makes sense, but I think a lot of the db structure and FK relationships revolve around those tables. I suspect if we get rid of them, you would also have to refactor out all mention of docid/rev in the whole system, which would be a monumental task.

**Comment by taojing2002** [comment]
Good point! That is a big task and we need to think about it.

**Comment by taojing2002** [comment]
I propose to drop this table as well:
```
metacat=# \d xml_accesssubtree 
                    Table "public.xml_accesssubtree"
    Column    |          Type          | Collation | Nullable | Default 
--------------+------------------------+-----------+----------+---------
 docid        | character varying(250) |           |          | 
 rev          | bigint                 |           |          | 1
 controllevel | character varying(50)  |           |          | 
 subtreeid    | character varying(250) |           |          | 
 startnodeid  | bigint                 |           |          | 
 endnodeid    | bigint                 |           |          | 
```
This table was used to support the feature in old version eml, which users can fine-tune the access control for parts of eml objects. Now the access control is totally based on system metadata for pids and we don't support the feature for long time. So we can drop this table as well.

**Comment by taojing2002** [comment]
@mbjones Currently, if Metacat can't fulfill a Metacat API request, Metacat will return a xml with the element `<error>`.  
Now, i am following this pattern to remove the Metacat API. If a client send a request via Metacat API action, such as `insert`, `query` and et al, it will get this message:
```
<error>This request is a part of the old Metacat API, which is no longer supported</error>
```

**Comment by mbjones** [comment]
That's good, @taojing2002 -- let's wordsmith a bit. How about:

```
<error>The original Metacat API has been replaced, and so this request is no longer supported. Equivalent API methods now are available through the DataONE API (see <https://hostname/context/api>).</error>
```

**Comment by taojing2002** [comment]
Sounds good. Thanks @mbjones

**Comment by taojing2002** [comment]
@mbjones I am planning to drop those tables:
```
xml_accesssubtree
xml_relation
xml_nodes
xml_nodes_revisions
xml_index
xml_returnfield
xml_queryresult
xml_path_index
scheduled_job
scheduled_job_params
```
The last two tables is used for scheduling Kepler web service jobs, I believe.

Also, I am considering to drop those three tables as well. They are related to LTER harvester:
```
harvest_site_schedule
harvest_log
harvest_detail_log
```

**Comment by mbjones** [comment]
Hi @taojing2002 Overall that sounds good. A few comments...

## Relationships

- `xml_relation`

Can you remind me what this table was for?

## Scheduled jobs
- `scheduled_job`
- `scheduled_job_params`

These two tables may be used for other scheduled tasks beyond Kepler. I think Mike Daigle developed a general scheduler approach with the objective of changing all of Metacat's timed tasks into ScheduledJob tasks, but I'm not sure how far he got in the conversion. I think it would be important to verify that nothing else uses this infrastructure beyond Kepler before removing it. And maybe think about whether the ScheduledJob pattern is a useful one to follow rather than using `Timer` directly.

## Harvester

- `harvest_site_schedule`
- `harvest_log`
- `harvest_detail_log`

These tables are not just for LTER, they are the basis for our OAI-PMH harvester, which certainly could have uses beyond LTER. We also expose OAI-PMH as a harvest protocol for external groups, and a coupl eof external sites use that to harvest our content into their systems. So, at a minimum, we should not remove the OAI-PMH protocol support in this refactor. I'm not sure how much the tables are involved in that PMH support.

**Comment by taojing2002** [comment]
@mbjones Okay, I will keep those tables:
```
scheduled_job
scheduled_job_params
harvest_site_schedule
harvest_log
harvest_detail_log
```

The xml_relations table stores the triple information for packaging based on docid. The structure looks like:
```
   Column    |          Type          | Collation | Nullable |                    Default                     
--------------+------------------------+-----------+----------+------------------------------------------------
 relationid   | bigint                 |           | not null | nextval('xml_relation_id_seq'::text::regclass)
 docid        | character varying(250) |           |          | 
 packagetype  | character varying(250) |           |          | 
 subject      | character varying(512) |           | not null | 
 subdoctype   | character varying(128) |           |          | 
 relationship | character varying(128) |           | not null | 
 object       | character varying(512) |           | not null | 
 objdoctype   | character varying(128) |           |          | 
```

Some examples are:
```
 44 | Rreeves.5                          | eml://ecoinformatics.org/eml-2.0.0                | Rreeves.5                          |            | Provides info for                                                                                      | Rreeves.4 
```
```
394 | ArchivalTag.5                      | -//ecoinformatics.org//eml-dataset-2.0.0beta6//EN | ArchivalTag.4    
                  |            | provides access control rules for                                                      
                | ArchivalTag.17
```

**Comment by taojing2002** [comment]
@mbjones Here are the tables we will drop:
```
- xml_accesssubtree
- xml_nodes
- xml_nodes_revisions
- xml_index
- xml_path_index
- xml_returnfield
- xml_queryresult
```
If you have any concern, please let know. Thanks.
